pkg/sdkstats/sdkstats.go

```diff
@@ -2,161 +2,123 @@ package sdkstats
-	SDKVersion   int
+	Version      string
 	CreatedAt    string
 	PushedAt     string
 	TotalCommits int
 }
 
-//GetStats returns []RepoMetaData populated with Code results from Github Search Code API, for a given search string
-func GetStats(client *github.Client, q string) ([]RepoMetadata, error) {
-	var repoList []RepoMetadata
-	var projectType string
-	var sdkVersion int
-
-	//Search function return CodeResults for the q queery, and also Basic repository data.
-	searchOp, err := Search(client, q)
-
-	if err != nil {
-		return nil, err
-	}
-
-	if strings.Contains(q, "helm") {
-		projectType = "helm"
-	}
-	if strings.Contains(q, "ansible") {
-		projectType = "ansible"
-	}
-	if strings.Contains(q, "go.mod") {
-		projectType = "go.mod"
-	}
-	if strings.Contains(q, "Gopkg.toml") {
-		projectType = "Gopkg.toml"
-	}
-	for j := 0; j < len(searchOp); j++ {
-		for i := 0; i < len(searchOp[j].CodeResults); i++ {
-			owner := searchOp[j].CodeResults[i].GetRepository().GetOwner().GetLogin()
-			name := searchOp[j].CodeResults[i].GetRepository().GetName()
-			fullName := searchOp[j].CodeResults[i].GetRepository().GetFullName()
+// RepoMetadataQuery struct provides search queries, projecttype, and VersionParser declaration
+type RepoMetadataQuery struct {
+	ProjectType   string
+	Queries       []string
+	VersionParser VersionParser
+}
 
-			//getVersion returns SDK version by doing Fragment search
-			if projectType == "Gopkg.toml" {
-				sdkVersion = 0
-			} else {
-				sdkVersion = getversion(searchOp[j].CodeResults[i], projectType)
-			}
+// VersionParser Interface implements ParseVersion(), to parse version number from a given Text Match result.
+type VersionParser interface {
+	ParseVersion(codeResults github.CodeResult) (string, error)
+}
 
-			//GetRepoDetails returns Repository specififc details
-			repoMap := GetRepoDetails(client, owner, name)
+// GetStats returns List of Repositories satisfyng the Search criteria,
+// populated with Stars, TotalCommits, Initial and Last Commit details.
+// Also included are basic details of repository such as URL, Owner, and name.
 
-			TotalCommits, err := strconv.Atoi(repoMap["TotalCommits"])
-			Stars, err := strconv.Atoi(repoMap["Stars"])
-			if err != nil {
-				return nil, err
-			}
-
-			repoDetails := &RepoMetadata{
-				FullName:     fullName,
-				ProjectType:  projectType,
-				SDKVersion:   sdkVersion,
-				Stars:        Stars,
-				CreatedAt:    repoMap["CreatedAt"],
-				PushedAt:     repoMap["PushedAt"],
-				TotalCommits: TotalCommits,
+func GetStats(ctx context.Context, tc *http.Client, rq RepoMetadataQuery) ([]RepoMetadata, error) {
+	var repoList []RepoMetadata
+	var repoDetails *RepoMetadata
+	for _, q := range rq.Queries {
+		searchOp, err := search(ctx, tc, q)
+		if err != nil {
+			return nil, err
+		}
+		for j := 0; j < len(searchOp); j++ {
+			for i := 0; i < len(searchOp[j].CodeResults); i++ {
+				repoDetails, err = getRepoDetails(ctx, tc, searchOp[j].CodeResults[i], rq)
```

Comment:
It looks like we're calling `getRepoDetails()` more than once per repository, which unnecessarily hits the Github API, driving up the query count and making things hit the rate limit faster.

We should refactor things so that we only call getRepoDetails once per repo.

Since a repo might have more than one operator built into it, maybe we call need `RepoMetadata` to have a `[]string` for `Versions`, and then we coul call `getRepoDetails()` once, and iterate the code results calling `rq.VersionParser.ParseVersion()` and appending to the `Versions` field.

WDYT?

Reply:
/hold